submissions/scripts/prep_files.py

```diff
@@ -0,0 +1,460 @@
+    external_metadata_samplify_fc_data['Investigator\'s sample ID'] = external_metadata_samplify_fc_data['Investigator\'s sample ID'].str.strip()
+    # only standardize internal barcodes
+    external_metadata_samplify_fc_data['Investigator\'s sample ID'] = external_metadata_samplify_fc_data.apply(lambda row: standardize_barcode(row['Investigator\'s sample ID']) if str(row['Project']).startswith(('starita_bbi_', 'SeattleChildrensDirect_')) else row['Investigator\'s sample ID'], axis=1)
+    external_metadata_samplify_fc_data['Collection date'] = external_metadata_samplify_fc_data['Collection date'].apply(standardize_date)
+
+    # Only the collection date from the Metadata sheet is used for creating submission files.
+    # Validating to confirm no collection dates prior to Feb 2020 and not all dates are identical.
+    validate_collection_dates(external_metadata_metadata['collection_date'])
+
+    sentinel_ids = ['blank', 'twist positive', 'pbs', 'lp_blank']
+    # all records with unique identifiers should be present in both sheets
+    non_matching_ids = pd.merge(external_metadata_metadata[~external_metadata_metadata["lab_accession_id"].str.lower().isin(sentinel_ids)],
+        external_metadata_samplify_fc_data[external_metadata_samplify_fc_data['Investigator\'s sample ID'].notna()],
+        how="outer",
+        left_on=["LIMS", "lab_accession_id"],
+        right_on=["Sample ID", "Investigator\'s sample ID"],
+        indicator=True).query("_merge != 'both'")[['LIMS', 'lab_accession_id', 'Sample ID',  'Investigator\'s sample ID']]
+
+    # for blanks and twist positives, the sample IDs should present on both sheets, but with empty `Investigator's sample ID` on Samplify FC Data sheet.
+    non_matching_blank_twist_pos = pd.merge(external_metadata_metadata[external_metadata_metadata["lab_accession_id"].str.lower().isin(sentinel_ids)],
+        external_metadata_samplify_fc_data[external_metadata_samplify_fc_data['Investigator\'s sample ID'].isna()],
+        how="outer",
+        left_on=["LIMS"],
+        right_on=["Sample ID"],
+        indicator=True).query("_merge != 'both'")[['LIMS', 'lab_accession_id', 'Sample ID',  'Investigator\'s sample ID']]
+
+    assert len(non_matching_ids) == 0, f"Error: Non-matching IDs found in metadata file:\n {non_matching_ids}"
+    assert len(non_matching_blank_twist_pos) == 0, f"Error: Non-matching blanks and/or twist positives found in metadata file:\n{non_matching_blank_twist_pos}"
+
+    # Identify and optionally remove expirimental samples
+    exp_samples = external_metadata_metadata[external_metadata_metadata['lab_accession_id'].str.endswith('_exp', na=False)]
+    exp_samples_samplify = external_metadata_samplify_fc_data[external_metadata_samplify_fc_data['Investigator\'s sample ID'].str.endswith('_exp', na=False)]
+    if not exp_samples.empty:
+        print(f"Expirimental samples found: {len(exp_samples)}\n {exp_samples.to_string()}")
+
+        if yes_no_cancel("Drop expirimental samples?"):
```

Comment:
Out of curiosity, are these experimental samples usually dropped?

Reply:
Yes, these are dropped after confirming with the sequencing team that they are cost center related. Otherwise, they may need to be included.